etl/open311/open311_to_socrata.py

```diff
@@ -0,0 +1,169 @@
+
+from etl import utils
+
+# Socrata Secrets
+SO_WEB = os.getenv("SO_WEB")
+SO_TOKEN = os.getenv("SO_TOKEN")
+SO_KEY = os.getenv("SO_KEY")
+SO_SECRET = os.getenv("SO_SECRET")
+REALTIME_DATASET = os.getenv("REALTIME_DATASET")
+
+# Open 311 API Key
+API_KEY = os.getenv("OPEN_311_API_KEY")
+API_BASE_URL = os.getenv("OPEN_311_API_BASE_URL")
+
+headers = {"Authorization": f"Bearer {API_KEY}"}
+
+# 10 requests per minute API limit
+MAX_REQUESTS_PER_MINUTE = 10
+SECONDS_PER_REQUEST = 60 / MAX_REQUESTS_PER_MINUTE
+
+# Timezone objects
+CENTRAL_TZ = ZoneInfo("America/Chicago")
+UTC_TZ = timezone.utc
+
+
+def determine_query_time(client, override_date=None):
+    """
+    Uses provided date arg first, and subtracts 10 minutes to reduce the risk we miss any updated records
+    If no arg is provided, the date used is the most recently updated socrata record.
+    If the dataset is empty, default to Oct 13, 2025.
+    """
+    if override_date:
+        try:
+            # Parse ISO 8601 datetime and subtract 10 minutes
+            dt = datetime.fromisoformat(override_date)
+            adjusted = dt - timedelta(minutes=10)
+            adjusted_str = adjusted.isoformat()
+            adjusted_str = adjusted_str.replace("+00:00", "Z")
+            logger.info(f"Using provided date (minus 10 minutes): {adjusted_str}")
+            return adjusted_str
+        except ValueError:
+            raise ValueError(
+                f"Invalid date format: {override_date}. Expected ISO 8601 format."
+            )
+
+    # get last updated record from socrata
+    date = client.get(
+        REALTIME_DATASET,
+        select="updated_datetime",
+        order="updated_datetime DESC",
+        limit=1,
+    )
+    if date:
+        # Interpret as Central Time and convert to UTC
+        ct_time = datetime.fromisoformat(date[0]["updated_datetime"]).replace(
+            tzinfo=CENTRAL_TZ
+        )
+        utc_time = ct_time.astimezone(UTC_TZ)
+        adjusted_str = utc_time.isoformat()
+        adjusted_str = adjusted_str.replace("+00:00", "Z")
+        return adjusted_str
+    return "2025-10-13T00:00:00Z"
+
+
+def socrata_point_location_formatting(rec):
+    """
+    Formats latitude and longitude for "Point" Socrata datatype
+    """
+    if rec["long"] and rec["lat"]:
+        return f"POINT ({rec['long']} {rec['lat']})"
+    return None
+
+
+def convert_to_central_and_strip_tz(rec, fields):
+    """
+    Converts UTC timestamps to Central Time and removes timezone info
+    (since Socrata stores floating datetimes in local time).
+    """
+    for f in fields:
+        if rec.get(f):
+            dt = datetime.fromisoformat(rec[f].replace("Z", "+00:00")).astimezone(
+                CENTRAL_TZ
+            )
+            rec[f] = dt.replace(tzinfo=None).isoformat()
+    return rec
+
+
+def main(args):
+    # Logging into socrata
+    soda = Socrata(
+        SO_WEB,
+        SO_TOKEN,
+        username=SO_KEY,
+        password=SO_SECRET,
+        timeout=500,
```

Comment:
8 and ⅓ minutes is a looooooong timeout. If socrata isn't answering us in 30 seconds, then I'd consider it not going to answer -- can we dial this way back? If not, we need to make sure we account for this timeout in the task and DAG timeouts and schedules in airflow.

Reply:
I'll dial this back but yeah if socrata is slow the DAG timeout will take over either way. We also have no way of knowing how many of these requests will be made to socrata as we're making a request every 100 records.